Implement EnumerateAudioDevices for FMOD

vfrz opened the pull request:
Next step would be to use FMOD callbacks to update device list (currently it is updating devices every update).

https://fmod.com/resources/documentation-api?version=2.0&page=core-api-system.html#system_setcallback
There is a specific callback for device list changes named `FMOD_SYSTEM_CALLBACK_DEVICELISTCHANGED`

Are you ok with this?

tlgkccampbell commented:
That sounds good to me. Do you plan to add that to this PR or do you want me to go ahead and merge it?

vfrz commented:
I can't get the callback working on my linux right now, seems it's an OS issue because even the getNumDrivers() remain the same while I plug/unplug different audio devices to my PC.
From what I've read from the FMOD forums, they are checking difference using getNumDrivers() to fire the callback so it's pretty much the same as what you've done in your C# implementation.

I will give it a try on a Windows machine and maybe contact the FMOD support to know a bit more about that.
I think you can keep the PR open for now, I'll let you know if I have any news.

vfrz commented:
I've implemented callback interop, that part is working fine.
I still get the problem about linux not detecting device changes but it doesn't break anything.

The PR is almost ready, I'll do some more tests on Windows tomorrow before merging.

I've removed all the `sealed` keywords in the two implementation classes since they are already sealed.

Out of subject question: is there any reason you are using the "capital named" primitive types? (Like `String` vs `string` or `Boolean` vs `bool`)

vfrz commented:
I've fixed the last bug I had about the GC collecting the callback.

PR is ready!

tlgkccampbell commented:
> Out of subject question: is there any reason you are using the "capital named" primitive types? (Like `String` vs `string` or `Boolean` vs `bool`)

I have Visual Studio configured to show reference types and value types in different colors, which helps me see potential issues with garbage collection. This doesn't work if you use the type keywords, because all keywords have the same color.

tlgkccampbell commented on Source/Ultraviolet.FMOD/Shared/Native/FMODNative.cs:
```diff
@@ -61,6 +61,9 @@ static FMODNative()
         
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static FMOD_RESULT FMOD_System_Update(FMOD_SYSTEM* system) => impl.FMOD_System_Update(system);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static FMOD_RESULT FMOD_System_SetCallback(FMOD_SYSTEM* system, FMOD_SYSTEM_CALLBACK callback, FMOD_SYSTEM_CALLBACK_TYPE callbackmask = FMOD_SYSTEM_CALLBACK_TYPE.ALL) => impl.FMOD_System_SetCallback(system, callback, callbackmask);
```
It looks like you modified `FMODNative` manually. Changes to the native API surface should be made by updating `_Definitionx.xml` and running the `UvNativeCodeGen` project, which will rebuild these files for you. This helps keep things in sync across the different platform implementations, and will become especially import when .NET 6 (theoretically) unifies all of this stuff, as I'll need to regenerate all of the native interface code to fit with whatever that ends up looking like.

vfrz replied:
Didn't know about that. I've pushed the fix!

tlgkccampbell commented:
This PR looks good to me with the exception noted above; if you can fix that, I'll merge it.
